### PR TITLE
Audit instance setup command integration test

### PR DESCRIPTION
**Pull request by cquirosj**

### CODE REVIEW

File: src/ServiceControl.Audit.AcceptanceTests/Auditing/When_running_setup_command__integration_test.cs [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+        public void Setup()
+        {
+            var transportIntegration = (ITransportIntegration)TestSuiteConstraints.Current.CreateTransportConfiguration();
+            Console.WriteLine($"\tConfigured transport configuration: {transportIntegration.Name}");
+            //Mark the test as inconclusive if when running in the CI SQL Server is not installed.
+#if !DEBUG
+            if (transportIntegration.Name.Equals("SQL Server", StringComparison.InvariantCultureIgnoreCase) == false)
+            {
+                _inconclusiveRun = true;
+                Assert.Inconclusive($"'");
```
Comment (timbussmann) [inline src/ServiceControl.Audit.AcceptanceTests/Auditing/When_running_setup_command__integration_test.cs:62]: can we provide a better message?

File: src/ServiceControl.Audit.AcceptanceTests/Auditing/When_running_setup_command__integration_test.cs [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+        public void TearDown()
+        {
+            if (_inconclusiveRun)
+            {
+                return;
+            }
+
+            _sqlLocalDbInstance?.Detach();
+            DeleteFolder(_workingDirectory);
+        }
```
Comment (timbussmann) [inline src/ServiceControl.Audit.AcceptanceTests/Auditing/When_running_setup_command__integration_test.cs:109]:
```suggestion
        }

```

File: src/ServiceControl.Audit.AcceptanceTests/Auditing/When_running_setup_command__integration_test.cs [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+            {
+                return;
+            }
+
+            try
+            {
+                emptyTempDirectory = new DirectoryInfo(Path.Combine(Path.GetTempPath(), Path.GetRandomFileName()));
+                emptyTempDirectory.Create();
+                var arguments = $"\"{emptyTempDirectory.FullName}\" \"{path.TrimEnd('\\')}\" /W:1  /R:1 /FFT /MIR /NFL";
+                using (var process = Process.Start(new ProcessStartInfo("robocopy")
```
Comment (timbussmann) [inline src/ServiceControl.Audit.AcceptanceTests/Auditing/When_running_setup_command__integration_test.cs:124]: what is this required for?

File: src/ServiceControl.Audit.AcceptanceTests/Auditing/When_running_setup_command__integration_test.cs [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+        }
+        #endregion TearDown
+
+        #region Test helpers
+        public class SqlLocalDb
+        {
+            public static SqlLocalDb CreateNewIn(string path)
+            {
+                return new SqlLocalDb(path);
+            }
```
Comment (timbussmann) [inline src/ServiceControl.Audit.AcceptanceTests/Auditing/When_running_setup_command__integration_test.cs:159]:
```suggestion
            }

```

File: src/ServiceControl.Audit.AcceptanceTests/Auditing/When_running_setup_command__integration_test.cs [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+
+                return (success: _installer.ReportCard.HasErrors == false, errors);
+            }
+
+            public string AuditQueueName => _installer.AuditQueue;
+            public string ErrorQueue => _installer.AuditQueue;
+        }
+        #endregion
+    }
+
```
Comment (timbussmann) [inline src/ServiceControl.Audit.AcceptanceTests/Auditing/When_running_setup_command__integration_test.cs:299]:
```suggestion
```

File: src/ServiceControl.Audit.AcceptanceTests/Auditing/When_running_setup_command__integration_test.cs [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+            AuditInstanceBinaries(TransportInfo transportInfo, string connectionString)
+            {
+                _connectionString = connectionString;
+                _transportInfo = transportInfo ?? throw new ArgumentNullException(nameof(transportInfo));
+            }
+
+            readonly TransportInfo _transportInfo;
+            readonly string _connectionString;
+        }
```
Comment (timbussmann) [inline src/ServiceControl.Audit.AcceptanceTests/Auditing/When_running_setup_command__integration_test.cs:276]:
```suggestion
        }

```

File: src/ServiceControl.Audit.AcceptanceTests/Auditing/When_running_setup_command__integration_test.cs [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+                using (var command = connection.CreateCommand())
+                {
+                    command.CommandText = sql;
+                    command.CommandType = CommandType.Text;
+                    return (int)command.ExecuteScalar();
+                }
+            }
+
+            readonly string _name;
+            public string ConnectionString => $@"Data Source=(LocalDB)\MSSQLLocalDB;Initial Catalog={_name};Integrated Security=True;Connection Timeout=300;";
```
Comment (timbussmann) [inline src/ServiceControl.Audit.AcceptanceTests/Auditing/When_running_setup_command__integration_test.cs:216]: isn't this the same as https://github.com/Particular/ServiceControl/pull/2886/files#diff-3bfe53cbb55e0ef07db3f50bc5e368875b362bc6a5e622c076d49889029d9dc3R192 ?

File: src/ServiceControl.Audit.AcceptanceTests/Auditing/When_running_setup_command__integration_test.cs [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+            public bool? HasTableWithName(string tableName)
+            {
+                using (var connection = NewConnection())
+                {
+                    connection.Open();
+                    var result = ExecuteScalar(connection, $"IF OBJECT_ID (N'[{tableName}]', N'U') IS NOT NULL SELECT 1 AS r ELSE SELECT 0 AS r;");
+                    return result == 1;
+                }
+            }
+        }
```
Comment (timbussmann) [inline src/ServiceControl.Audit.AcceptanceTests/Auditing/When_running_setup_command__integration_test.cs:227]:
```suggestion
        }

```

File: src/ServiceControl.Audit.AcceptanceTests/Auditing/When_running_setup_command__integration_test.cs (hunk 3 above)
Comment (cquirosj) [reply src/ServiceControl.Audit.AcceptanceTests/Auditing/When_running_setup_command__integration_test.cs:124]: At first, I tried to just use Directory.Delete but it failed. I copied this code over from the Acceptance Tests base classes that actually also rely on invoking the SetupBootstraper

File: src/ServiceControl.Audit.AcceptanceTests/Auditing/When_running_setup_command__integration_test.cs [hunk 9]
```diff
@@ -0,0 +1,300 @@
...
+    using System.IO;
+    using System.Linq;
+    using System.Security.AccessControl;
+    using System.Security.Principal;
+    using AcceptanceTesting;
+
+    [TestFixture]
+    public class When_running_setup_command__integration_test
+    {
+        string _workingDirectory;
```
[the remark this answers is not in the document]
Comment (cquirosj) [reply src/ServiceControl.Audit.AcceptanceTests/Auditing/When_running_setup_command__integration_test.cs:21]: It is an old habit, I will fix it.

File: src/ServiceControl.Audit.AcceptanceTests/Auditing/When_running_setup_command__integration_test.cs (hunk 1 above)
Comment (cquirosj) [reply src/ServiceControl.Audit.AcceptanceTests/Auditing/When_running_setup_command__integration_test.cs]: Good catch

File: src/ServiceControl.Audit.AcceptanceTests/Setup/DirectoryDeleter.cs [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+{
+    using System.Diagnostics;
+    using System.IO;
+    using System.Linq;
+    using System.Security.AccessControl;
+    using System.Security.Principal;
+
+    static class DirectoryDeleter
+    {
+        public static void Delete(string path)
```
Comment (SzymonPobiega) [inline src/ServiceControl.Audit.AcceptanceTests/Setup/DirectoryDeleter.cs:11]: Could you add a comment here describing why this is needed vs using APIs? That could be helpful for the next person that works on these tests :)

**Review by SzymonPobiega — APPROVED** [review]
Awesome. I really like that this part is not tested ❤️ I left just one suggestion about a comment that might be helpful for our future selves.

**Comment by cquirosj** [comment]
@timbussmann please re-review. I did some refactorings to the test helpers, removed the regions, and re-organized the files. I think it reads better now.

File: src/ServiceControl.Audit.AcceptanceTests/Setup/DirectoryDeleter.cs (hunk 10 above)
Comment (cquirosj) [reply src/ServiceControl.Audit.AcceptanceTests/Setup/DirectoryDeleter.cs:16]: done.

File: src/ServiceControl.Audit.AcceptanceTests/Setup/DirectoryDeleter.cs [hunk 11]
```diff
@@ -0,0 +1,58 @@
+﻿namespace ServiceControl.Audit.AcceptanceTests.Setup
+{
+    using System.Diagnostics;
+    using System.IO;
+    using System.Linq;
+    using System.Security.AccessControl;
+    using System.Security.Principal;
+
+    static class DirectoryDeleter
```
Comment (timbussmann) [inline src/ServiceControl.Audit.AcceptanceTests/Setup/DirectoryDeleter.cs:9]: should this helper method also be used in other places where this approach is used? (e.g. https://github.com/Particular/ServiceControl/blob/c1a58beef3a405209c2664186da83e611574b614/src/ServiceControl.Audit.AcceptanceTests/TestSupport/ServiceControlComponentRunner.cs )

File: src/ServiceControl.Audit.AcceptanceTests/Setup/When_running_setup_command__integration_test.cs [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+    using System;
+    using Infrastructure.Installers;
+    using NUnit.Framework;
+    using System.Diagnostics;
+    using System.IO;
+    using System.Linq;
+    using AcceptanceTesting;
+
+    [TestFixture]
+    public class When_running_setup_command__integration_test
```
Comment (timbussmann) [inline src/ServiceControl.Audit.AcceptanceTests/Setup/When_running_setup_command__integration_test.cs:12]:
```suggestion
    public class When_running_setup_command_integration_test
```

**Review by SzymonPobiega — APPROVED** [review]

File: src/ServiceControl.Audit.AcceptanceTests/Setup/DirectoryDeleter.cs (hunk 11 above)
Comment (cquirosj) [reply src/ServiceControl.Audit.AcceptanceTests/Setup/DirectoryDeleter.cs:9]: Yeah, the code is exactly the same. I will push the change and ensure tests continue to pass.

**Pull request merged**